jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java

```diff
@@ -0,0 +1,366 @@
+    /**
+     * Map with light edges for each vertex. An edge is considered
+     * light if its weight is less than or equal to {@link #delta}.
+     */
+    private Map<V, Set<E>> light;
+    /**
+     * Map with light edges for each vertex. An edge is
+     * considered heavy if its weight is greater than {@link #delta}.
+     */
+    private Map<V, Set<E>> heavy;
+
+    /**
+     * Map that stores information about each vertex.
+     *
+     * <p>
+     * In each triple the first value stands for the bucket index of a
+     * vertex or $-1$ if a vertex does not belong to any bucket. The second
+     * value stands for the tentative distance to a vertex. The third value
+     * of each triple stands for the predecessor of a vertex in the the
+     * shortest path tree. The second and the third values of each triple will
+     * be used at the end of the computation to construct shortest paths tree.
+     *
+     * <p>
+     * Keeping vertex information in an {@link AtomicReference} objects allows
+     * to avoid threads synchronisation. Thus a thread can safely update
+     * the information using standard CAS function.
+     */
+    private Map<V, AtomicReference<Triple<Integer, Double, E>>> verticesDataMap;
+
+    /**
+     * Constructs a new instance of the algorithm for a given graph.
+     * Initializes {@link #delta} to $0.0$ to preserve lazy computation stype.
+     *
+     * @param graph graph
+     */
+    public DeltaSteppingShortestPath(Graph<V, E> graph) {
+        super(graph);
+        delta = 0.0;
+        initializeFields();
+    }
+
+    /**
+     * Constructs a new instance of the algorithm for a given graph and delta.
+     *
+     * @param graph the graph
+     * @param delta bucket width
+     */
+    public DeltaSteppingShortestPath(Graph<V, E> graph, double delta) {
+        super(graph);
+        if (delta <= 0) {
+            throw new IllegalArgumentException(DELTA_MUST_BE_POSITIVE);
+        }
+        this.delta = delta;
+        initializeFields();
+    }
+
+    /**
+     * Initializes {@link #light}, {@link #heavy} and {@link #verticesDataMap} fields.
+     */
+    private void initializeFields() {
+        light = new HashMap<>();
+        heavy = new HashMap<>();
+        verticesDataMap = new HashMap<>();
+    }
+
+    /**
+     * Asserts that all edges in the {@link #graph} have positive weights.
+     */
+    private void assertPositiveWeights() {
+        boolean allEdgesWithNonNegativeWeights = graph.edgeSet().stream()
+                .map(graph::getEdgeWeight).allMatch(weight -> weight >= 0);
+        if (!allEdgesWithNonNegativeWeights) {
+            throw new IllegalArgumentException(NEGATIVE_EDGE_WEIGHT_NOT_ALLOWED);
+        }
+    }
+
+    /**
+     * Calculates max edge weight in the {@link #graph}.
+     *
+     * @return max edge weight
+     */
+    private Optional<Double> maxEdgeWeight() {
+        return graph.edgeSet().stream().map(graph::getEdgeWeight).max(Double::compare);
```

Comment:
Indeed. It can be eliminated by replacing `.max(Double::compare)` with `.mapToDouble(d -> d).max()`. However this then requires that the result type be `OptionalDouble`, which takes less space but has less methods.